internal/pkg/service/common/configmap/visit.go

```diff
@@ -0,0 +1,167 @@
+	"fmt"
+	"github.com/keboola/go-utils/pkg/orderedmap"
+	"reflect"
+)
+
+type VisitConfig struct {
+	OnField func(field reflect.StructField) (fieldName string, ok bool)
+	OnValue func(vc *VisitContext) error
+}
+
+type VisitContext struct {
+	// StructField contains metadata about the structure field, for example tags.
+	StructField reflect.StructField
+	// OriginalPath is path to the filed in the source structure.
+	OriginalPath orderedmap.Path
+	// MappedPath is path to the field modified by the VisitConfig.OnField function.
+	MappedPath orderedmap.Path
+	// Value is untouched Value of the structure field.
+	Value reflect.Value
+	// PrimitiveValue is Value converted to base/primitive type, if it is possible,
+	// otherwise the PrimitiveValue is same as the Value.
+	PrimitiveValue reflect.Value
+	// Type of the Value
+	Type      reflect.Type
+	Sensitive bool
+	Usage     string
+}
+
+func Visit(value reflect.Value, cfg VisitConfig) error {
+	vc := &VisitContext{}
+	vc.Value = value
+	vc.PrimitiveValue = value
+	vc.Type = value.Type()
+	return doVisit(vc, cfg)
+}
+
+func doVisit(vc *VisitContext, cfg VisitConfig) error {
+	isNil := vc.Value.Kind() == reflect.Pointer && vc.Value.IsNil()
+
+	onPrimitiveValue := func(primitiveValue reflect.Value) error {
+		vc.PrimitiveValue = primitiveValue
+		return cfg.OnValue(vc)
+	}
+
+	// Dereference pointer
+	originalValue := vc.Value
+	for vc.Value.Kind() == reflect.Pointer && !vc.Value.IsNil() {
+		vc.Value = vc.Value.Elem()
+		vc.Type = vc.Type.Elem()
+	}
+
+	// Convert type to a pointer, unmarshal methods may use pointers in method receivers
+	methodReceiver := originalValue
+	if originalValue.Kind() != reflect.Pointer {
+		ptr := reflect.New(originalValue.Type())
+		ptr.Elem().Set(originalValue)
+		methodReceiver = ptr
+	}
+
+	// Check if the struct implements an unmarshaler
+	switch v := methodReceiver.Interface().(type) {
+	case fmt.Stringer:
+		if isNil {
+			return onPrimitiveValue(reflect.ValueOf(""))
+		}
+		return onPrimitiveValue(reflect.ValueOf(v.String()))
+	case json.Marshaler:
+		if isNil {
+			return onPrimitiveValue(reflect.ValueOf(""))
+		}
+		if v, err := v.MarshalJSON(); err == nil {
+			return onPrimitiveValue(reflect.ValueOf(string(v)))
+		} else {
+			return err
+		}
+	case encoding.TextMarshaler:
+		if isNil {
+			return onPrimitiveValue(reflect.ValueOf(""))
+		}
+		if v, err := v.MarshalText(); err == nil {
+			return onPrimitiveValue(reflect.ValueOf(string(v)))
+		} else {
+			return err
+		}
+	case encoding.BinaryMarshaler:
+		if isNil {
+			return onPrimitiveValue(reflect.ValueOf(""))
+		}
+		if bytes, err := v.MarshalBinary(); err == nil {
+			return onPrimitiveValue(reflect.ValueOf(string(bytes)))
+		} else {
+			return err
+		}
+	default:
+		switch vc.Value.Kind() {
+		case reflect.Int:
+			return onPrimitiveValue(reflect.ValueOf(int(vc.Value.Int())))
+		case reflect.Int8:
+			return onPrimitiveValue(reflect.ValueOf(int8(vc.Value.Int())))
+		case reflect.Int16:
```

Comment:
Toto sa zial neda zapisat kratsie.

`.Kind()` je primitivny typ, ktorý stojí za daným typom.
- Pre `type GzipLevel int`, to bude `int`.